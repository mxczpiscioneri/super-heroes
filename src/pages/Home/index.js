import React, { Component } from 'react'
import { FlatList } from 'react-native'
import PropTypes from 'prop-types'
import { Background, Character } from '../../components'
import { generateUrlImage } from '../../utils'
import { getCharacters } from '../../services/api'

class Home extends Component {
  constructor(props) {
    super(props)

    this.state = {
      characters: [],
      refreshing: false,
      page: 0,
    }
  }

  componentDidMount() {
    this.loadCharacters()
  }

  /**
   * Method to load characters
   * @param {String} params - Options for filters
   */
  async loadCharacters(params = '') {
    this.setState({ refreshing: true })

    const limitCharacters = 20
    const response = await getCharacters(`?limit=${limitCharacters}&offset=${this.state.page * limitCharacters}${params}`)

    const characters = [...this.state.characters, ...response.results]

    this.setState({
      characters,
      refreshing: false
    })
  }

  /**
   * Method to load more characters
   */
  handleLoadMore() {
    this.setState(prevState => ({
      page: prevState.page + 1,
    }), () => {
      this.loadCharacters()
    })
  }

  render() {
    const { navigate } = this.props.navigation

    return (
      <Background>
        <FlatList
          data={this.state.characters}
          keyExtractor={item => `characters${item.id}`}
          numColumns={2}
          refreshing={this.state.refreshing}
          onRefresh={() => this.loadCharacters()}
          onEndReached={() => this.handleLoadMore()}
          onEndThreshold={0.5}
          onEndReachedThreshold={0.5}
          renderItem={
            ({ item }) => (
              <Character
                character={item}
                legend
                image={generateUrlImage(item.thumbnail.path, item.thumbnail.extension, 'standard_fantastic')}
                onPress={() => navigate('Details', { characterId: item.id })}
              />
            )
          }
        />
      </Background>
    )
  }
}

Home.propTypes = {
  navigation: PropTypes.object.isRequired,
}

export default Home
