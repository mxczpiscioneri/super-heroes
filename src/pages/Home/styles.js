import Styled from 'styled-components/native'

export const Icon = Styled.TouchableOpacity`
  padding: 12px;
  paddingRight: 16px;
`

export const Image = Styled.Image`
  
`
