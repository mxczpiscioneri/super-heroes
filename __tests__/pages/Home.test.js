
import React from 'react'
import renderer from 'react-test-renderer'
import Home from '../../src/pages/Home'

const navigation = { navigate: jest.fn() }

it('Home snapshot', () => {
  const rendered = renderer.create(
    <Home navigation={navigation} />
  ).toJSON()
  expect(rendered).toMatchSnapshot()
})

it('Home snapshot', () => {
  const rendered = renderer.create(
    <Home navigation={navigation} />
  ).toJSON()
  expect(rendered).toMatchSnapshot()
})
