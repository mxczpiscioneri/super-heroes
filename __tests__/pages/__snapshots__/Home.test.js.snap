// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home snapshot 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/statics/images/background.jpg",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={2}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onEndThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={true}
      />
    }
    refreshing={true}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;

exports[`Home snapshot 2`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/statics/images/background.jpg",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={2}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onEndThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={true}
      />
    }
    refreshing={true}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;
